File: dotnet/examples/Monitoring/MissingTopicNotifications.cs
/**
 * Copyright © 2023 - 2024 Diffusion Data Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Generic;
using static System.Console;
using System.Linq;
using PushTechnology.ClientInterface.Client.Callbacks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features;
using PushTechnology.ClientInterface.Client.Features.Control.Topics;
using PushTechnology.ClientInterface.Client.Features.Topics;
using PushTechnology.ClientInterface.Client.Session;
using PushTechnology.ClientInterface.Client.Topics;
using PushTechnology.ClientInterface.Client.Topics.Details;
using static PushTechnology.ClientInterface.Examples.Program;
using NUnit.Framework;

namespace PushTechnology.ClientInterface.Examples.Monitoring
{
    public sealed class MissingTopicNotifications : Example
    {
        public override async Task Run(CancellationToken cancellationToken, string[] args)
        {
            string serverUrl = args[0];

            var session = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            IRegistration registration = null;
            string topicPath = "my/topic/path";
            string topicSelector = "?my/topic/path//";

            var missingTopicNotificationStream = new MissingTopicNotificationStream(session);
            registration = await session.TopicControl.AddMissingTopicHandlerAsync(topicPath, missingTopicNotificationStream, cancellationToken);

            var session2 = Diffusion.Sessions
                .Principal("client")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            var stringStream = new StringStream();
            session2.Topics.AddStream(topicSelector, stringStream);

            await session2.Topics.SubscribeAsync("my/topic/path/does/not/exist/yet", cancellationToken);

            session.Close();
            session2.Close();
        }

        private sealed class MissingTopicNotificationStream : IMissingTopicNotificationStream
        {
            private ISession session;
            public MissingTopicNotificationStream(ISession session) => this.session = session;

            public void OnClose() {}

            public void OnError(ErrorReason errorReason) {}

            public void OnMissingTopic(IMissingTopicNotification notification)
            {
                WriteLine($"Topic '{notification.TopicPath}' does not exist.");

                session.TopicControl.AddTopicAsync(notification.TopicPath, session.TopicControl.NewSpecification(TopicType.STRING));
            }
        }

        private sealed class StringStream : IValueStream<string>
        {
            public void OnClose() {}

            public void OnError(ErrorReason errorReason) {}

            public void OnSubscription(string topicPath, ITopicSpecification specification)
            {
                WriteLine($"Subscribed to {topicPath}.");
            }

            public void OnUnsubscription(string topicPath, ITopicSpecification specification, TopicUnsubscribeReason reason)
            {
                WriteLine($"Unsubscribed from {topicPath}: {reason}.");
            }

            public void OnValue(string topicPath, ITopicSpecification specification, string oldValue, string newValue)
            {
                WriteLine($"{topicPath} changed from {oldValue ?? "NULL"} to {newValue}.");
            }
        }
    }
}

File: c/examples/monitoring/missing_topic_notifications.c
/**
 * Copyright © 2024 DiffusionData Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This example is written in C99. Please use an appropriate C99 capable compiler
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "diffusion.h"
#include "utils.h"
MUTEX_DEF

SESSION_T *g_control_session;

static int on_error(
    SESSION_T *session,
    const DIFFUSION_ERROR_T *error)
{
    printf("On error: %s\n", error->message);
    return HANDLER_SUCCESS;
}

static int on_missing_topic(
    SESSION_T * session,
    const SVC_MISSING_TOPIC_REQUEST_T *request,
    void *context)
{
    printf("Received missing topic notification:\n");

    char *session_id = session_id_to_string(request->session_id);
    printf("\tSession ID: %s\n", session_id);
    free(session_id);

    printf("\tTopic Selector: %s\n", request->topic_selector);

    utils_create_string_topic(
        g_control_session, "my/topic/path/does/not/exist/yet", "Hello");

    MUTEX_BROADCAST
    return HANDLER_SUCCESS;
}


void run_example(
    const char *url,
    const char *principal,
    CREDENTIALS_T *credentials)
{
    MUTEX_INIT

    SESSION_T *admin_session = utils_open_session(url, "admin", "password");

    const char *topic_path_root = "my/topic/path";
    MISSING_TOPIC_PARAMS_T missing_topic_params = {

        .topic_path = topic_path_root,
        .on_missing_topic = on_missing_topic,
        .on_error = on_error
    };
    CONVERSATION_ID_T *missing_topic_cid =
        missing_topic_register_handler(admin_session, missing_topic_params);

    g_control_session = utils_open_session(url, "control", "password");

    VALUE_STREAM_T *value_stream_ptr = utils_subscribe(
        g_control_session, "my/topic/path/does/not/exist/yet", DATATYPE_STRING
    );
    MUTEX_WAIT

    // Sleep for a bit to see the notifications
    sleep(2);

    missing_topic_deregister_handler(admin_session, missing_topic_cid);

    // Sleep for a bit to see the notifications
    sleep(2);

    session_close(g_control_session, NULL);
    session_free(g_control_session);
    free(value_stream_ptr);

    session_close(admin_session, NULL);
    session_free(admin_session);

    MUTEX_TERMINATE
}